Application for plotting array

VictorBarbosaMartins opened the pull request:
Closes #339

Implemented the application for plotting the array. For more examples please see the integration tests.
This PR is built upon #428. Please approve and merge that one first.

GernotMaier commented:
Did you also update the documentation? Better to check the docu how to use the application then the integration tests...

VictorBarbosaMartins commented:
> Did you also update the documentation? Better to check the docu how to use the application then the integration tests...

I just updated the documentation as well, thanks for the reminder. 

I implemented the application from scratch, there was no documentation. The tests were failing due to a feature in the code, they should pass now.

Before reviewing here, we need to approve and merge #428 . There we haven't agreed yet on passing 'site' as a parameter to LayoutArray. Orel and myself are for passing it. I suggest we approve there and if later we decide not to pass 'site' anymore we adapt the code but it is important now for the further development to have it.

orelgueta commented on applications/plot_layout_array.py:
```diff
@@ -0,0 +1,182 @@
+    config.parser.add_argument(
+        "--rotate_angle",
+        help="Angle to rotate the array before plotting (in degrees).",
+        type=str,
```
Why is this a string and not a float?

VictorBarbosaMartins replied:
I had some troubles with initializing it as None (and type float) so I defined it to be a str. would it be better to define it as float and initialize as 0.0?

orelgueta commented on applications/plot_layout_array.py:
```diff
@@ -0,0 +1,182 @@
+        default=None,
+    )
+
+    return config.initialize(db_config=True)
```
Why is the db_config set to True here? I do not see the DB actually used below.

VictorBarbosaMartins replied:
I have copied it from the other application but If we dont need it I am removing the dependence now

orelgueta commented on applications/plot_layout_array.py:
```diff
@@ -0,0 +1,182 @@
+        rotate_angles = [0 * u.deg]
+    else:
+        rotate_angles = []
+        for _, one_angle in enumerate(args_dict["rotate_angle"]):
```
Why do you use `enumerate` without actually using the counter? Seems unnecessary.

VictorBarbosaMartins replied:
I thought codacy would complain but you are right, since I am using only the value and not the counter I dont need enumerate. I am removing it now

orelgueta commented on applications/plot_layout_array.py:
```diff
@@ -0,0 +1,182 @@
+            io_handler_instance.get_input_data_file(
+                "layout", f"telescope_positions-{one_array}.ecsv"
+            )
+            for _, one_array in enumerate(args_dict["layout_array_name"])
```
Same question as above about `enumerate`

orelgueta commented on applications/plot_layout_array.py:
```diff
@@ -0,0 +1,182 @@
+    .. code-block:: console
+
+        python applications/plot_layout_array.py --figure_name northern_array_alpha \
+        --layout_array_name test_layout
```
I modified this to be a working example

VictorBarbosaMartins replied:
I dont understand. What would be a working example? And have you pushed any changes to the PR ? I don't see your commit

orelgueta replied:
Sorry, I committed but forgot to push...
See now the changes.

orelgueta commented on applications/plot_layout_array.py:
```diff
@@ -0,0 +1,182 @@
+            )
+            plot_file = output_dir.joinpath(plot_file_name)
+
+            for f in ["pdf", "png"]:
```
I modified this to be `ext`, but it's not important.

VictorBarbosaMartins replied:
again, I dont see your commit. please remember to checkout the branch after my changes in case you want to change directly something in the code

orelgueta replied:
Again, forgot to push. This resulted in a conflict so I reimplemented the small change after your modifications. Your modifications are a bit more complex than I expected though, so I will review that now.

orelgueta reviewed (commented):
A few minor comments and made some very minor changes. 

However, the real issue is when I looked at the plots. They are rotated with respect to reality. The south is rotated by 90 degrees, the north I think as well, but I am not sure. Also, the SSTs are way too small and are essentially invisible with this colour. 

Also, when plotting just 4 LSTs with `--layout_array_name North-4LST`, the plot is cut and therefore shows only two LSTs.

VictorBarbosaMartins commented on applications/plot_layout_array.py:
```diff
@@ -0,0 +1,182 @@
+    Command line arguments
+    ----------------------
+    figure_name (str, optional)
+        File name for the pdf output (without extension).
```
Indeed it would be better. I implemented my idea, please take a look it thats fine for you.

orelgueta replied:
Yes, the current implementation is fine, resolving this thread.

VictorBarbosaMartins commented:
@orelgueta thanks for the comments. Some points are still open, and I am still going to check if the arrays are rotate and why. I am still going to adjust the size of the sst markers as you suggested.

VictorBarbosaMartins commented:
> However, the real issue is when I looked at the plots. They are rotated with respect to reality. The south is rotated by 90 degrees, the north I think as well, but I am not sure. Also, the SSTs are way too small and are essentially invisible with this colour.
I have checked and the arrays are not rotated. You can confirm it by checking the plotted image with the values in the ecsv file. Here I am assuming that east is the positive X axis and north is the positive Y axis. If this is not the case, please let me know.

VictorBarbosaMartins commented:
> Also, when plotting just 4 LSTs with `--layout_array_name North-4LST`, the plot is cut and therefore shows only two LSTs.
The plot is not cut but if you look at `data/layout/telescope_positions-4LST.ecsv` you will see that two telescopes are placed at the same position twice. I don't actually know why this was done (I havent created those files) or whether it makes sense, but the problem here is not with the plotting. Please let me know if the positions of telescopes have to be changed in the file.

VictorBarbosaMartins commented:
> Also, the SSTs are way too small and are essentially invisible with this colour.

I returned to the definition of a scale parameter to increase the size of the markers in case the total number of telescopes is larger than 30 and changed the color of the SST marker. Furthermore, I increased the dpi of the saved image, which is necessary in case we want to zoom in.

Printing the label of the telescopes is still not ideal, but it will hardly be, since it is hard to plot 100 markers with a label in a reasonable size without overlaying each other (Try `python applications/plot_layout_array.py --telescope_list data/layout/telescope_positions-South-TestLayout.ecsv --figure_name test.png --show_tel_label` and look at `simtools-output/plot_array_layout/applications-plots/test.png` to see what I mean)

orelgueta commented:
> > However, the real issue is when I looked at the plots. They are rotated with respect to reality. The south is rotated by 90 degrees, the north I think as well, but I am not sure. Also, the SSTs are way too small and are essentially invisible with this colour.
> 
> I have checked and the arrays are not rotated. You can confirm it by checking the plotted image with the values in the ecsv file. Here I am assuming that east is the positive X axis and north is the positive Y axis. If this is not the case, please let me know.

The coordinate system of the CORSIKA inputs is what's rotated I think. We'll confirm and continue discussing this on Mattermost.

orelgueta commented:
> > Also, when plotting just 4 LSTs with `--layout_array_name North-4LST`, the plot is cut and therefore shows only two LSTs.
> 
> The plot is not cut but if you look at `data/layout/telescope_positions-4LST.ecsv` you will see that two telescopes are placed at the same position twice. I don't actually know why this was done (I havent created those files) or whether it makes sense, but the problem here is not with the plotting. Please let me know if the positions of telescopes have to be changed in the file.

Yes, I noticed it now when reviewing the other PR. That is clearly a typo in the telescope positions which should be fixed in those files.

orelgueta commented:
> > Also, the SSTs are way too small and are essentially invisible with this colour.
> 
> I returned to the definition of a scale parameter to increase the size of the markers in case the total number of telescopes is larger than 30 and changed the color of the SST marker. Furthermore, I increased the dpi of the saved image, which is necessary in case we want to zoom in.
> 
> Printing the label of the telescopes is still not ideal, but it will hardly be, since it is hard to plot 100 markers with a label in a reasonable size without overlaying each other (Try `python applications/plot_layout_array.py --telescope_list data/layout/telescope_positions-South-TestLayout.ecsv --figure_name test.png --show_tel_label` and look at `simtools-output/plot_array_layout/applications-plots/test.png` to see what I mean)

The plot looks good now. The SSTs are much more visible and even when printing the labels everything is fairly clear. It might be even clearer after rotating the plot.

orelgueta reviewed (approved):
Approving after the various changes implemented and discussion on Mattermost
